lxd/locking/lock.go

```diff
@@ -16,47 +16,67 @@ var locksMutex sync.Mutex
+// On success, it returns an unlock function which needs to be called to unlock the lock and the channel which can be used to fait for operation to finish.
+// On failure, it returns only channel.
+func lock(lockName string) (UnlockFunc, chan struct{}) {
+	// Get exclusive access to the map and see if there is already an operation ongoing.
+	locksMutex.Lock()
+	waitCh, ok := locks[lockName]
+
+	if !ok {
+		// No ongoing operation, create a new channel to indicate our new operation.
+		waitCh = make(chan struct{})
+		locks[lockName] = waitCh
+		locksMutex.Unlock()
+
+		// Return a function that will complete the operation.
+		return func() {
+			// Get exclusive access to the map.
+			locksMutex.Lock()
+			doneCh, ok := locks[lockName]
 
-			// Return a function that will complete the operation.
-			return func() {
-				// Get exclusive access to the map.
-				locksMutex.Lock()
-				doneCh, ok := locks[lockName]
-
-				// Load our existing operation.
-				if ok {
-					// Close the channel to indicate to other waiting users
-					// they can now try again to create a new operation.
-					close(doneCh)
-
-					// Remove our existing operation entry from the map.
-					delete(locks, lockName)
-				}
-
-				// Release the lock now that the done channel is closed and the
-				// map entry has been deleted, this will allow any waiting users
-				// to try and get access to the map to create a new operation.
-				locksMutex.Unlock()
+			// Load our existing operation.
+			if ok {
+				// Close the channel to indicate to other waiting users
+				// they can now try again to create a new operation.
+				close(doneCh)
+
+				// Remove our existing operation entry from the map.
+				delete(locks, lockName)
 			}
-		}
 
-		// An existing operation is ongoing, lets wait for that to finish and then try
-		// to get exlusive access to create a new operation again.
-		locksMutex.Unlock()
+			// Release the lock now that the done channel is closed and the
+			// map entry has been deleted, this will allow any waiting users
+			// to try and get access to the map to create a new operation.
+			locksMutex.Unlock()
+		}, waitCh
+	}
+
+	// An existing operation is ongoing, lets wait for that to finish and then try
+	// to get exlusive access to create a new operation again.
+	locksMutex.Unlock()
+
+	return nil, waitCh
+}
+
+// Lock creates the lock for storage volume.
+// Will block until the lock is established. On success, it returns an unlock function which needs to be called to unlock the lock.
+func Lock(lockName string) UnlockFunc {
+	for {
+		unlock, waitCh := lock(lockName)
+		if unlock != nil {
+			return unlock
+		}
 		<-waitCh
 	}
 }
+
+// TryLock creates the lock for storage volume.
+// It's not blocking function. On success, it returns an unlock function which needs to be called to unlock the lock.
+// On failure, it returns nil.
+func TryLock(lockName string) UnlockFunc {
```

Comment:
We probably dont need these changes, see my other comments below.

Reply:
I've removed this part.